Clean up docs to with regards to changes in SELFDESTRUCT opcode (#14803)
ethereum/solidity: docs/introduction-to-smart-contracts.rst

contracts, the Ether is forever lost.

.. warning::

>     From ``EVM >= Cancun`` onwards, ``selfdestruct`` will **only** send all Ether in the account to the given recipient. In the case when ``selfdestruct`` is called in the same transaction the contract was created in, the behaviour of ``selfdestruct`` described above the warning pre-Cancun hardfork (i.e., ``EVM <= Shanghai``) is preserved and will destroy the current contract, deleting any data, including storage keys, code and the account itself.

Reviewer: What about:
```suggestion
    From ``EVM >= Cancun`` onwards, ``selfdestruct`` will **only** send all Ether in the account to the given recipient and not destroy the contract. However, when ``selfdestruct`` is called in the same transaction as the contract calling ``selfdestruct`` was created, the behaviour of ``selfdestruct`` pre-Cancun hardfork (i.e., ``EVM <= Shanghai``) is preserved and will destroy the current contract, deleting any data, including storage keys, code and the account itself. See `EIP-6780 <https://eips.ethereum.org/EIPS/eip-6780>`_ for more details.
```

>     Also, from version 0.8.18 and up, the use of ``selfdestruct`` in both Solidity and Yul will trigger a
>     deprecation warning, since the ``SELFDESTRUCT`` opcode will eventually undergo breaking changes in behavior
>     as stated in `EIP-6049 <https://eips.ethereum.org/EIPS/eip-6049>`_.

Reviewer: Maybe rephrase this to something like:
```suggestion
    Also, note that the ``selfdestruct`` opcode remains deprecated since Solidity version 0.8.18, and the use of the opcode in both Solidity and Yul will trigger a deprecation warning, since the opcode is expected to undergo substantial behavioral changes, as stated in `EIP-6049 <https://eips.ethereum.org/EIPS/eip-6049>`_.
    Moreover, future adjustments to the EVM may further restrict the functionality of this opcode.
```

Author: Yes, this makes more sense. Thank you! :)

contracts, the Ether is forever lost.

.. warning::

>     From ``EVM >= Cancun`` onwards, ``selfdestruct`` will **only** send all Ether in the account to the given recipient and not destroy the contract. However, when ``selfdestruct`` is called in the same transaction as the contract calling ``selfdestruct`` was created, the behaviour of ``selfdestruct`` pre-Cancun hardfork (i.e., ``EVM <= Shanghai``) is preserved and will destroy the current contract, deleting any data, including storage keys, code and the account itself. See `EIP-6780 <https://eips.ethereum.org/EIPS/eip-6780>`_ for more details.

Reviewer 2: This bit sounds a bit awkward:
```suggestion
    From ``EVM >= Cancun`` onwards, ``selfdestruct`` will **only** send all Ether in the account to the given recipient and not destroy the contract. However, when ``selfdestruct`` is called in the same transaction that creates the contract calling it, the behaviour of ``selfdestruct`` before Cancun hardfork (i.e., ``EVM <= Shanghai``) is preserved and will destroy the current contract, deleting any data, including storage keys, code and the account itself. See `EIP-6780 <https://eips.ethereum.org/EIPS/eip-6780>`_ for more details.
```

> 

Reviewer 2: One thing that's still missing is https://github.com/ethereum/solidity/pull/14803#discussion_r1511340392. I.e. I think it's important to mention the distinction between EVM version used by the network and the EVM version we're compiling for.

We could use the wording based on the blog post:
```
This is a network-wide change that affected all contracts present on the Ethereum mainnet and testnets.
The ``--evm-version`` setting used when compiling the contract has no bearing on it.
```

Reviewer 2: I see you added this to one of the other warnings. I think it's related to the one here instead so I just pushed a commit that moved it here.

    Also, note that the ``selfdestruct`` opcode remains deprecated since Solidity version 0.8.18, and the use of the opcode in both Solidity and Yul will trigger a deprecation warning, since the opcode is expected to undergo substantial behavioral changes, as stated in `EIP-6049 <https://eips.ethereum.org/EIPS/eip-6049>`_.

>     Moreover, future adjustments to the EVM may further restrict the functionality of this opcode.

Reviewer 2: Just a general remark - note that the docs break long lines. This makes them generally easier to edit and to read in the source form. The line breaks disappear when the document is rendered.

I'd particularly recommend breaking the line after each sentence (or earlier if it's still too long), because this way you don't have to redo the line breaks in the whole paragraph if you change it. It also makes the diff more readable (only the changed sentence is highlighted). It also makes it much easier to suggest short corrections on github - see in particular my next comment - I had to include the whole paragraph, but it would be much clearer if could comment just on the part I'm correcting.

Here's how I'd do this for this bit:

```suggestion
    From ``EVM >= Cancun`` onwards, ``selfdestruct`` will **only** send all Ether in the account to the given recipient and not destroy the contract.
    However, when ``selfdestruct`` is called in the same transaction as the contract calling ``selfdestruct`` was created,
    the behaviour of ``selfdestruct`` pre-Cancun hardfork (i.e., ``EVM <= Shanghai``) is preserved and will destroy the current contract,
    deleting any data, including storage keys, code and the account itself.
    See `EIP-6780 <https://eips.ethereum.org/EIPS/eip-6780>`_ for more details.

    Also, note that the ``selfdestruct`` opcode remains deprecated since Solidity version 0.8.18, and the use of the opcode
    in both Solidity and Yul will trigger a deprecation warning, since the opcode is expected to undergo substantial behavioral changes,
    as stated in `EIP-6049 <https://eips.ethereum.org/EIPS/eip-6049>`_.
    Moreover, future adjustments to the EVM may further restrict the functionality of this opcode.
```

Author: Thanks for the pointers, @cameel! I will keep this in mind.

>     Also, note that the ``selfdestruct`` opcode remains deprecated since Solidity version 0.8.18, and the use of the opcode in both Solidity and Yul will trigger a deprecation warning, since the opcode is expected to undergo substantial behavioral changes, as stated in `EIP-6049 <https://eips.ethereum.org/EIPS/eip-6049>`_.
>     Moreover, future adjustments to the EVM may further restrict the functionality of this opcode.

Reviewer 2: Well, it has already undergone substantial behavioral changes :)

How about using the wording from the blog post?
```suggestion
    Also, note that the ``selfdestruct`` opcode has been deprecated in Solidity version 0.8.18, as recommended by `EIP-6049 <https://eips.ethereum.org/EIPS/eip-6049>`_.
    The deprecation is still in effect and the compiler will still emit warnings on its use.
    Any use in newly deployed contracts is strongly discouraged even if the new behavior is taken into account.
    Future changes to the EVM might further reduce the functionality of the opcode.
```

    However, when ``selfdestruct`` is called in the same transaction that creates the contract calling it,
    the behaviour of ``selfdestruct`` before Cancun hardfork (i.e., ``EVM <= Shanghai``) is preserved and will destroy the current contract,
    deleting any data, including storage keys, code and the account itself.
    See `EIP-6780 <https://eips.ethereum.org/EIPS/eip-6780>`_ for more details.

    The new behaviour is the result of a network-wide change that affects all contracts present on
    the Ethereum mainnet and testnets.
    It is important to note that this change is dependent on the EVM version of the chain on which
    the contract is deployed.
    The ``--evm-version`` setting used when compiling the contract has no bearing on it.

    Also, note that the ``selfdestruct`` opcode has been deprecated in Solidity version 0.8.18,
    as recommended by `EIP-6049 <https://eips.ethereum.org/EIPS/eip-6049>`_.
    The deprecation is still in effect and the compiler will still emit warnings on its use.
    Any use in newly deployed contracts is strongly discouraged even if the new behavior is taken into account.

>     Future changes to the EVM might further reduce the functionality of the opcode.

Reviewer 2: @mehtavishwa30 Taking a final look at the rendered version, this whole section about `selfdestruct` actually strikes me as too advanced for "Introduction to Smart Contracts". Mentioning `selfdestruct` made sense when it was in common use, but now that it has all these warnings, it's no longer recommended for newbies to touch it. Maybe we should simplify it for new users?

We could delete everything from it and leave only a short paragraph saying that contracts cannot be deleted, only deactivated. We still have a copy of most of these warnings in "Units and Global Variables" so we could link there for anyone who wants the details. But in the intro we'd only be mentioning the fact that something called `selfdestruct` exists, but you should no be tempted to use it for deleting contracts since it's deprecated or something like that.

I mean, this is not necessary in this PR, I'm merging it as is, but could be done in a follow-up PR if you want to brush up the intro page a little.
